Arctan avx512

Ka-zam opened the pull request:
Added AVX512 kernels and some minor cleanup.

Using AVX512F yields 40% speedup over the AVX2_FMA implementation on my 7950X3D. Compared to the generic atan2 implementation this is a 65x speedup.

```
                      magnus@r7950x3d:~/src/kazam/volk/build$ volk_profile -R atan
                      RUN_VOLK_TESTS: volk_32f_atan_32f(131071,1987)
                      generic completed in 2010.97 ms
                      polynomial completed in 62.8946 ms
                      a_avx512 completed in 40.8423 ms
                      a_avx2_fma completed in 56.9026 ms
                      a_avx2 completed in 55.9691 ms
                      a_sse4_1 completed in 110.292 ms
                      u_avx512 completed in 39.5152 ms
                      u_avx2_fma completed in 55.4739 ms
                      u_avx2 completed in 55.6364 ms
                      u_sse4_1 completed in 110.009 ms
                      Best aligned arch: u_avx512
                      Best unaligned arch: u_avx512
                      RUN_VOLK_TESTS: volk_32fc_s32f_atan2_32f(131071,1987)
 ------>              generic completed in 4199.28 ms
                      polynomial completed in 95.92 ms
 ------>              a_avx512 completed in 64.0566 ms
                      a_avx2_fma completed in 99.0502 ms
                      a_avx2 completed in 98.3313 ms
 ------>              u_avx512 completed in 63.4753 ms
                      u_avx2_fma completed in 98.3834 ms
                      u_avx2 completed in 98.6633 ms
                      Best aligned arch: u_avx512
                      Best unaligned arch: u_avx512
                      Writing /home/magnus/.volk/volk_config...
```

Ka-zam committed:
Includes for GCC 7.2 compatibility

Ka-zam committed:
Merge branch 'master' of github.com:gnuradio/volk

Ka-zam committed:
Merge branch 'master' of github.com:Ka-zam/volk

Ka-zam committed:
Merge branch 'master' of github.com:Ka-zam/volk

jdemel committed:
gtest: Start work on new test infrastructure

Use GTest to run tests. This should help us in quite a few places.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

Ka-zam committed:
Merge branch 'gnuradio:master' into master

jdemel committed:
ci: Add a second test example

This is an ongoing effort to introduce some more sophisticated QA tests.
We use googletest and write custom test suits.

This approach requires more manual work but yields a more accurate
result as well.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
tests: Add a log info print test

Potentially, we want to be more verbose. This is a first effort to
improve in this area.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
cpu_features: Update hints in README

We should prefer a cpu_features version that is provided thru package
managers. The README clearly states this preference now.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

Ka-zam committed:
Added AVX512 kernels to arctan and atan2

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Removed comment, simplified

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Restored file

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
New AVX512F implementation

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Updated copyright year

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam commented:
I just noticed there's a NaN test as well... 

https://github.com/gnuradio/volk/pull/731

Need to update this PR with this as well for AVX512!

jj1bdx commented:
@Ka-zam #731 is essential for airspy-fmradion, and I've spent a few weeks solving the NaN issue. Please add the NaN test before completing your implementation.

jdemel commented on include/volk/volk_avx512_intrinsics.h:
```diff
@@ -0,0 +1,64 @@
+ * Maximum relative error ~6.5e-7
+ * Polynomial evaluated via Horner's method
+ */
+static inline __m512 _mm512_arctan_poly_avx512(const __m512 x)
+{
+    const __m512 a1 = _mm512_set1_ps(+0x1.ffffeap-1f);
+    const __m512 a3 = _mm512_set1_ps(-0x1.55437p-2f);
```
Since AVX512 comes in a variety of flavors, it might be worth to add a comment in every function which extensions need to be present. Although, I assume in this case it's just foundation.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+static inline void
+volk_32f_atan_32f_generic(float* out, const float* in, unsigned int num_points)
+{
+    unsigned int number = 0;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+volk_32f_atan_32f_polynomial(float* out, const float* in, unsigned int num_points)
+{
+    unsigned int number = 0;
+    for (; number < num_points; number++) {
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+    const __m512 abs_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x7FFFFFFF));
+    const __m512 sign_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x80000000));
+
+    unsigned int number = 0;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+        out += 16;
+    }
+
+    number = sixteenth_points * 16;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -100,6 +100,66 @@ static inline void volk_32fc_s32f_atan2_32f_polynomial(float* outputVector,
+        __m512 input = _mm512_div_ps(_mm512_mask_blend_ps(swap_mask, y, x),
+                                     _mm512_mask_blend_ps(swap_mask, x, y));
+        __mmask16 nan_mask = _mm512_cmp_ps_mask(input, input, _CMP_UNORD_Q);
+        input = _mm512_mask_blend_ps(nan_mask, input, zero);
```
So you already included the NaN check?

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+            _mm512_cmp_ps_mask(_mm512_and_ps(x, abs_mask), one, _CMP_GT_OS);
+        __m512 x_star = _mm512_div_ps(_mm512_mask_blend_ps(swap_mask, x, one),
+                                      _mm512_mask_blend_ps(swap_mask, one, x));
+        __m512 result = _mm512_arctan_poly_avx512(x_star);
```
The NaN check needs to be added here too.

jdemel replied:
Since this is the `atan` function, not to be confused with the `atan2` function, I assume, we can avoid the NaN check.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+    const __m512 abs_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x7FFFFFFF));
+    const __m512 sign_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x80000000));
+
+    unsigned int number = 0;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+        out += 16;
+    }
+
+    number = sixteenth_points * 16;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+    const __m512 sign_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x80000000));
+
+    unsigned int number = 0;
+    unsigned int sixteenth_points = num_points / 16;
```
should be `const`. At least that would make reading the code easier.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -100,6 +100,66 @@ static inline void volk_32fc_s32f_atan2_32f_polynomial(float* outputVector,
+    const __m512 sign_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x80000000));
+    const __m512 zero = _mm512_setzero_ps();
+
+    unsigned int number = 0;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -100,6 +100,66 @@ static inline void volk_32fc_s32f_atan2_32f_polynomial(float* outputVector,
+    const __m512 zero = _mm512_setzero_ps();
+
+    unsigned int number = 0;
+    unsigned int sixteenth_points = num_points / 16;
```
Better make this a `const` variable.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+            _mm512_cmp_ps_mask(_mm512_and_ps(x, abs_mask), one, _CMP_GT_OS);
+        __m512 x_star = _mm512_div_ps(_mm512_mask_blend_ps(swap_mask, x, one),
+                                      _mm512_mask_blend_ps(swap_mask, one, x));
+        __m512 result = _mm512_arctan_poly_avx512(x_star);
```
I guess, the NaN check is missing here.

Ka-zam replied:
The Nan check is strictly for atan2, right?

jdemel replied:
we use `atanf` in `generic`. Thus, we should check against this functions' behavior: https://en.cppreference.com/w/c/numeric/math/atan

I couldn't spot an issue while reading the code. Should be fine.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -100,6 +100,66 @@ static inline void volk_32fc_s32f_atan2_32f_polynomial(float* outputVector,
+
+    number = sixteenth_points * 16;
+    volk_32fc_s32f_atan2_32f_polynomial(
+        out, (lv_32fc_t*)in, normalizeFactor, num_points - number);
```
why is this cast required here? It might be better to just do `complexVector + number`.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -224,6 +284,66 @@ static inline void volk_32fc_s32f_atan2_32f_a_avx2(float* outputVector,
+    const __m512 sign_mask = _mm512_castsi512_ps(_mm512_set1_epi32(0x80000000));
+    const __m512 zero = _mm512_setzero_ps();
+
+    unsigned int number = 0;
```
please move this into the for loop.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -224,6 +284,66 @@ static inline void volk_32fc_s32f_atan2_32f_a_avx2(float* outputVector,
+    number = sixteenth_points * 16;
+    volk_32fc_s32f_atan2_32f_polynomial(
+        out, (lv_32fc_t*)in, normalizeFactor, num_points - number);
+}
```
my comment in the aligned kernel applies here as well.

jdemel reviewed (changes requested):
Thanks for your contribution. The `NaN` checks should really be integrated. Other than that, only minor comments.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -59,6 +59,64 @@
+#include <immintrin.h>
+#include <volk/volk_avx512_intrinsics.h>
+static inline void
+volk_32f_atan_32f_a_avx512(float* out, const float* in, unsigned int num_points)
```
It would make sense to call this kernel `avx512dq` because this distinguishes it from just the foundation, aka `F` AVX512 stuff.

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+#include <immintrin.h>
+#include <volk/volk_avx512_intrinsics.h>
+static inline void
+volk_32f_atan_32f_u_avx512(float* out, const float* in, unsigned int num_points)
```
`_avx512dq`?

jdemel commented on kernels/volk/volk_32f_atan_32f.h:
```diff
@@ -168,6 +226,42 @@ volk_32f_atan_32f_a_sse4_1(float* out, const float* in, unsigned int num_points)
+        *out++ = volk_arctan(*in++);
+    }
+}
+#endif /* LV_HAVE_AVX512F for unaligned */
```
`LV_HAVE_AVX512DQ ...`?

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -100,6 +100,66 @@ static inline void volk_32fc_s32f_atan2_32f_polynomial(float* outputVector,
+#if LV_HAVE_AVX512F && LV_HAVE_AVX512DQ
+#include <immintrin.h>
+#include <volk/volk_avx512_intrinsics.h>
+static inline void volk_32fc_s32f_atan2_32f_a_avx512(float* outputVector,
```
`dq` should probably be part of the implementation name.

jdemel commented on kernels/volk/volk_32fc_s32f_atan2_32f.h:
```diff
@@ -224,6 +284,66 @@ static inline void volk_32fc_s32f_atan2_32f_a_avx2(float* outputVector,
+#if LV_HAVE_AVX512F && LV_HAVE_AVX512DQ
+#include <immintrin.h>
+#include <volk/volk_avx512_intrinsics.h>
+static inline void volk_32fc_s32f_atan2_32f_u_avx512(float* outputVector,
```
`dq`?

Ka-zam commented:
> @Ka-zam #731 is essential for airspy-fmradion, and I've spent a few weeks solving the NaN issue. Please add the NaN test before completing your implementation.

I think it's already in there and should work fine! Wrote a test program and
``` 
atan2(0.f, 0.f) == 0.f
```
 for all kernels.

Ka-zam committed:
resolved comments

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
formatting

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

jdemel committed:
cmake: Fix 64bit host CPU detection

In cases where we don't cross-compile, we might want to detect if a CPU
is 32bit or 64bit. CMake provides functionality for this case starting
in CMake 3.10. Let's use it.
Ubuntu 20.04 uses CMake 3.16. From the top of my head, this is the
oldest supported distribution. Debian buster ships with CMake 3.13.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

Ka-zam commented:
Here's some special values and a sanity check:

```
magnus@r7950x3d:~/src/kazam/scratch$ ./a.out 
          y :  1.00  -1.00   1.00  -1.00    nan    nan   0.00  -0.00   1.00  -1.00 
          x :  1.00   1.00  -1.00  -1.00   1.00    nan   0.00   0.00   0.00   0.00 
 atan2(y, x):
    generic :  0.79  -0.79   2.36  -2.36    nan    nan   0.00  -0.00   1.57  -1.57 
 polynomial :  0.79  -0.79   2.36  -2.36    nan    nan   0.00  -0.00   1.57  -1.57 
 a_avx512dq :  0.79  -0.79   2.36  -2.36   0.00   0.00   0.00   0.00   1.57  -1.57 
```

Do we care about the sign of atan2(-0, 0)? What about propagating nan?

jdemel committed:
macos: Fix CI dependency error

The following error popped up during our CI builds on MacOS:
```shell
$ pip3 install mako
  shell: /bin/bash -e {0}
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-brew-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.

    If you wish to install a non-brew packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```
[PEP668](https://peps.python.org/pep-0668/) tries to help people to
prevent dependency issues on their systems.
We can be pretty sure that our controlled CI environment is not affected
by this issue. Thus, we can use `--break-system-packages` and install
system-wide.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

drmpeg committed:
cmake: Suppress invalid escape sequence warnings with Python 3.12

Signed-off-by: Ron Economos <w6rz@comcast.net>

doronbehar committed:
cmake/pkgconfig: use CMAKE_INSTALL_FULL_* variables

See: https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html

Signed-off-by: Doron Behar <doron.behar@gmail.com>

jdemel committed:
gtest: Make gtest an install dependency

It's probably a good thing to have googletest as an install dependency.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
gtest: Enable GTests in CI workflows

We want to gradually migrate to a more stable test framework. This
requires that our CI runs the new framework. Here we go!

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
tests: Beautify test output

We can run an individual test for each implementation now. Further, the
test names are easier to recognize.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

camel-cdr committed:
add RISC-V Vector extension (RVV) kernels

Signed-off-by: Olaf Bernstein <camel-cdr@protonmail.com>

jdemel committed:
ci: Fix run-on-arch hiccup

We probably need to load a newer version than what we pinned before.
Let's just update and hopefully enjoy the benefits.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

camel-cdr committed:
add braces around one-liners for loops

Signed-off-by: Olaf Bernstein <camel-cdr@protonmail.com>

camel-cdr committed:
use with submodules recursive in RVV CI

Signed-off-by: Olaf Bernstein <camel-cdr@protonmail.com>

camel-cdr committed:
RVV CI: build sequentially with verbose

Signed-off-by: Olaf Bernstein <camel-cdr@protonmail.com>

camel-cdr committed:
use segmented store in rotator2 rvvseg implementation

Signed-off-by: Olaf Bernstein <camel-cdr@protonmail.com>

Zaryob committed:
Changes that suggested on #770

Writing to the python helper script moved to `python/volk_modtool/CMakeLists.txt`
Revert "Changes that suggested on #770" and Added way given on pull req.

This reverts commit d4afa9ee1edbab2b4c737db288359e339d5a00a9 partially.

as @jdemel stated:
> CMAKE_CURRENT_BINARY_DIR would possibly be the preferred way to use. I'd suspect changing this variable in the original location would already fix your issue. That'd be great. That way we can keep the code in a more generic place that makes it easier to re-use. Also, thanks for your patience.

I suspected about is there any upper config deletes CMAKE_CURRENT_SOURCE_DIR and CMAKE_CURRENT_BUILD_DIR and it overriding on 'CMakeLists.txt'.

Signed-off-by: Suleyman Poyraz <zaryob.dev@gmail.com>

tinyboxvk committed:
appveyor: Update to VS 2022/Python 3.12

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

tinyboxvk committed:
Update android_build.yml

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

tinyboxvk committed:
Adds toolchain file for Raspberry Pi 5

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

tinyboxvk committed:
Update links

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

tinyboxvk committed:
Update links

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

tinyboxvk committed:
Fix clang-format issue

Signed-off-by: tinyboxvk <13696594+tinyboxvk@users.noreply.github.com>

jdemel committed:
cleanup: Remove unused includes

clangd reports multiple unused includes. These should be removed to
potentially reduce build times and also to minimize includes. Only
including necessary includes helps to better understand the code.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
ci: Add first native Linux ARM runners

Since there are native ARM runners available now, we should use them.
Let's test this.

This is the source: https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories
As indicated, this is still in beta.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

SJ-Innovation committed:
feature: add env variable kernel override

Signed-off-by: Sam Lane <sl01172@surrey.ac.uk>

jsallay committed:
Add const to several args

Signed-off-by: John Sallay <jasallay@gmail.com>

jsallay committed:
Run clang-format

Signed-off-by: John Sallay <jasallay@gmail.com>

SJ-Innovation committed:
fix: formatting

Signed-off-by: Sam Lane <sl01172@surrey.ac.uk>

jdemel committed:
zenodo: Update citations

With new contributors come new entries in the Zenodo file. Further, git
supports .mailmap to canonicalize all the different names and email
addresses that people use over the years.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
Release 3.2.0

anilgurses committed:
Fix github actions for publishing docs

Signed-off-by: Anıl Gürses <anilgurses98@gmail.com>

marcusmueller committed:
Add tool to compare profiles

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

jdemel committed:
readme: Add paragraph to outline dependency policy

This should help users and contributors to better understand the
different needs/desires. A written statement can be discussed.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
ci: Remove obsolete CI, add new CI

This PR switches to later Ubuntu versions, and runs the CI with later
compilers.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
filesystem: Fully rely on std::filesystem

In the past, we tried to detect different options for std::filesystem,
or even used Boost. Now is a good time to remove all the old code and
just rely on the fact that current compilers provide std::filesystem
with C++17.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
ci: Add Werror flag to CI for C compilation

We only used the Werror flag for C++. If we use the Werror flag, we
should mostly use it for our core.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
cx-limited-range: Scope compile feature

We only want to use this flag in necessary compilation units. All other
units should build without this option. Further, this fixes issues with
the latest Clang compiler versions.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
cpu_features: Start using packaged version in CI

Most distros provide a cpu_features package now. On most platforms all
the required features are available. Let's use these features.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
ppc64le: Suppress unused variable warning

The cpu_features build does not use some variables for the ppc64le
build. This errors with the Werror flag. This commit suppresses this
error.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
cmake: Fix auto-format

We use cmake-format to have consistent CMake files.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
neon: Fix compile check

The `check_c_source_compiles` code to test if NEON is supported was
broken. Let's add flags and formatting to make it work again.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
16u_bytswap: Update code style

Move loop variable declaration into loop. Fix `maybe-uninitialized`
warning by initializing the corresponding variable.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
cmake: Update to modern PIC enablement

Instead of adding a `-fPIC` flag, we can use more modern CMake features
to enable `POSITION INDEPENDENT CODE` (PIC).

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

jdemel committed:
rotator: Add specialized gtest suite

This test suite should enable us to test the rotator more specifically.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>

Ka-zam committed:
Merge remote-tracking branch 'origin/master' into arctan_avx512

Ka-zam committed:
Add edge cases and tests to atan2/atan

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Merge branch 'main' of github.com:gnuradio/volk into arctan_avx512

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Merge main branch into arctan_avx512

Ka-zam committed:
removed dupes

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Fix log2 edge case handling for NaN and negative inputs

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Fix log2 edge case handling for NaN and negative inputs

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
added riscv kernels

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
fix RVV atan2 nan propagation

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
use vmandnot instead of vmandn

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
revert...

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Add infinity handling to RVV atan2 implementation

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Fix RVV acos/asin: one_minus_v_sq intrinsic

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
asin: use correct formula (inverse of acos)

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
use HUGE_VALF instead of __builtin_inff for portability

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

Ka-zam committed:
Add DCO signoff for commit faf20a4

I hereby add my Signed-off-by to this commit: faf20a4
Includes for GCC 7.2 compatibility

This commit was made before DCO was required for this project.
I certify that I have the rights to submit this work under the
project's license.

Signed-off-by: Magnus Lundmark <magnuslundmark@gmail.com>

jdemel commented:
I just merged your other PR. I assume this one is obsolete now. Thank you!
